src/v/cluster/archival/tests/archival_metadata_stm_test.cc

```diff
@@ -985,3 +985,105 @@ FIXTURE_TEST(test_highest_producer_id, archival_metadata_stm_fixture) {
     BOOST_REQUIRE_EQUAL(
       archival_stm->manifest().highest_producer_id(), model::producer_id{100});
 }
+
+FIXTURE_TEST(
+  test_archival_stm_spillover_failure, archival_metadata_stm_fixture) {
+    wait_for_confirmed_leader();
+
+    // Metadata with inconsistencies (gap)
+    std::vector<cloud_storage::segment_meta> m;
+    m.push_back(segment_meta{
+      .size_bytes = 100,
+      .base_offset = model::offset(0),
+      .committed_offset = model::offset(990),
+      .delta_offset = model::offset_delta(0),
+      .archiver_term = model::term_id(1),
+      .segment_term = model::term_id(1)});
+    m.push_back(segment_meta{
+      .size_bytes = 200,
+      .base_offset = model::offset(1000),
+      .committed_offset = model::offset(1999),
+      .archiver_term = model::term_id(1),
+      .segment_term = model::term_id(1)});
+
+    auto batcher = archival_stm->batch_start(
+      ss::lowres_clock::now() + 10s, never_abort);
+    batcher.add_segments(m, cluster::segment_validated::yes);
+    batcher.replicate().get();
+    BOOST_REQUIRE_EQUAL(archival_stm->manifest().size(), 2);
+    BOOST_REQUIRE_EQUAL(archival_stm->get_start_offset(), model::offset(0));
+    // Archive not initialized yet
+    BOOST_REQUIRE_EQUAL(
+      archival_stm->manifest().get_archive_start_offset(), model::offset());
+
+    // Set up empty archive
+    archival_stm
+      ->truncate_archive_init(
+        model::offset{0},
+        model::offset_delta(0),
+        ss::lowres_clock::now() + 10s,
+        never_abort)
+      .get();
+    BOOST_REQUIRE_EQUAL(
+      archival_stm->manifest().get_archive_start_offset(), model::offset(0));
+    BOOST_REQUIRE_EQUAL(
+      archival_stm->manifest().get_archive_clean_offset(), model::offset());
+    BOOST_REQUIRE_EQUAL(
+      archival_stm->manifest().full_log_start_kafka_offset(), kafka::offset(0));
+
+    archival_stm
+      ->cleanup_archive(
+        model::offset{0}, 0, ss::lowres_clock::now() + 10s, never_abort)
+      .get();
+    BOOST_REQUIRE_EQUAL(
+      archival_stm->manifest().get_archive_start_offset(), model::offset{});
+    BOOST_REQUIRE_EQUAL(
+      archival_stm->manifest().get_archive_clean_offset(), model::offset{});
+
+    vlog(
+      logger.info,
+      "start offset: {}",
+      archival_stm->manifest().get_start_offset());
+    vlog(
+      logger.info,
+      "kafka start offset: {}",
+      archival_stm->manifest().full_log_start_kafka_offset());
+    vlog(
+      logger.info,
+      "archive start offset: {}",
+      archival_stm->manifest().get_archive_start_offset());
+    vlog(
+      logger.info,
+      "archive clean offset: {}",
+      archival_stm->manifest().get_archive_clean_offset());
+
+    // The spillover command is misaligned because of the gap.
```

Comment:
I'm looking at the code and I see truncate_archive_init being called with spillover base offset:

```
const auto first = *tail.begin();
const auto spillover_meta = tail.make_manifest_metadata();
batch.spillover(spillover_meta);
batch.truncate_archive_init(first.base_offset, first.delta_offset);
```

Where the difference in offset 0 and 100 comes from 🤔

Reply:
100 is arbitrary, the goal here is for this value to be larger than 0 (current start offset) so it could trigger the assertion when the commands are applied incorrectly

Reply:
What I don't get is why you chose value of 100 when the way I'm reading the code this is an impossible scenario because `batch.truncate_archive_init` is called with `first.base_offset()` which in this test can't be anything else than 0 (base offset of the first segment).

Reply:
Actually not even this. Retention would move _start_offset and archival will match it.

I can't think of a realistic scenario which could lead to this case... :/